% Source: https://software.broadinstitute.org/gatk/best-practices/

\documentclass{beamer}
\usepackage{graphicx}
\newcommand {\framedgraphic}[1] {
\begin{center}
\includegraphics[width=\textwidth,height=0.8\textheight,keepaspectratio]{#1}
\end{center}
}
\AtBeginSection[]
{
\begin{frame}
\frametitle{Table of Contents}
\tableofcontents[currentsection]
\end{frame}
}
\usetheme{Warsaw}

\title{GATK Variant Call Best Practices}
\subtitle{BROAD Institute}
\author{Emil V}
\institute{Institute of Cytology \& Genetics, SB RAS}
\date{\today}

\begin{document}

\begin{frame}
\titlepage
\end{frame}

\section{Data pre-processing for variant discovery}

\subsection{Map to Reference}

\begin{frame}
\frametitle{Map to Reference}

\begin{enumerate}
\item BWA --- mapping
\begin{itemize}
\item Good at mapping chimera reads
\item Main alignment is soft-clipped, supplementary alignments are hard-clipped
\item Multi-threading support 
\end{itemize}
\item MergeBamAlignments (Picard) --- merging files
\end{enumerate}

\end{frame}

\begin{frame}
\frametitle{Alignment Types}
\framedgraphic{cigar.png}
\end{frame}

\subsection{Mark Duplicates \& Sort}

\begin{frame}
\frametitle{Mark Duplicates \& Sort}

\begin{enumerate}
\item MarkDuplicatesSpark (GATK)
\begin{itemize}
\item Perform both the duplicate marking step and the sort step
\item Multi-threading support
\item Can be run locally, without Spark cluster
\item Machines with a high number of cores or fast disk drives
\end{itemize}
\item MarkDuplicates + SortSam (Picard)
\begin{itemize}
\item Duplicate marking and sorting by coordinate are divided
\item Single-threaded
\end{itemize}
\end{enumerate}

\begin{block}{Note}
Both tools require query-sorted input. 
\end{block}

\end{frame}

\begin{frame}
\frametitle{How Picard MarkDuplicates Works}
\begin{enumerate}
\item Finds 5' coordinates of each read, taking into account clipping, gaps, or jumps.
\item Matches all pairs using 5' coordinates and orientation
\item Marks all pairs but the one with the highest sum of base qualities
\end{enumerate}

\begin{block}{Note}
With query-sorted input, if canonical read (or pair) is marked as dup, supplementary and secondary reads of the query will be marked as well.
With coordinate-sorted input, supplementary and secondary reads will be ignored.
\end{block}

\end{frame}

\subsection{Base (Quality Score) Recalibration (BQSR)}

\begin{frame}
\frametitle{Base (Quality Score) Recalibration (BQSR)}

GATK tools:

\begin{enumerate}
\item BaseRecalibrator: build ML model on dataset
\item Apply Recalibration: apply the model
\item AnalyzeCovariates: evaluate and compare base quality score
\end{enumerate}

\end{frame}

\begin{frame}
\frametitle{BQSR Details}

ML model is built on the following features:
\begin{itemize}
\item Read group the read belongs to
\item Quality score reported by the machine
\item Machine cycle producing this base (base number from the start of the read)
\item Current base + previous base (dinucleotide)
\end{itemize}
Loci known to vary in the population (typically found in dbSNP) are excluded from training data.

\end{frame}

\begin{frame}
\frametitle{BQSR Success Factors}

\begin{enumerate}
\item Read Group tags --- \texttt{@RG}
\item Amount of data (more than 100M bases per RG)
\item Model trained on your own variants dataset (for non-human data)
\end{enumerate}

\begin{alertblock}{Important}
GATK strongly recommends to perform BQSR on your data, \textit{almost} always.
\end{alertblock}

\end{frame}

\begin{frame}
\frametitle{BQSR Results}
\framedgraphic{qual.png}
\end{frame}

\section{Short variant discovery}

\subsection{Germline variants in single-sample mode}

\begin{frame}
\frametitle{Germline variant discovery in single-sample mode}
\framedgraphic{single_pipeline.png}
\end{frame}

\subsection{Call variants per-sample}

\begin{frame}
\frametitle{Call variants per-sample}

\begin{itemize}
\item Freebayes + vcfallelicprimitives
\item HaplotypeCaller (GATK) in single-sample mode
\end{itemize}

\end{frame}

\subsection{Callset refinement}

\begin{frame}
\frametitle{Callset refinement}

\begin{enumerate}
\item CNNScoreVariants: annotates each variant with a score indicating the model's prediction of the quality of each variant
\item FIlterVariantTranches: apply filters based on calculated quality score and variant sensitivity tranches appropriate for the task
\end{enumerate}

\end{frame}

\subsection{Germline Cohort Analysis}

\begin{frame}
\frametitle{Germline Cohort Analysis}
\framedgraphic{cohort_pipeline.png}
\end{frame}

\subsection{Somatic short variants}

\begin{frame}
\frametitle{Somatic short variants}
\framedgraphic{somatic_pipeline.png}
\end{frame}

\begin{frame}
\frametitle{Germline pipeline comparison}
\begin{itemize}
 \item Mutect, unlike HaplotypeCaller, discards existing mapping information and completely reassembles the reads in that region in order to generate candidate variant haplotypes
 \item CalculateContamination: estimate of the allelic copy number segmentation of each tumor sample
 \item LearnReadOrientationModel: learn the parameters of a model for orientation bias. This is extremely important for FFPE tumor samples.
\end{itemize}

\end{frame}

\section{Annotation}

\begin{frame}
\frametitle{Tools}

\begin{enumerate}
\item Ensembl VEP
\item Annovar
\item Funcotator (GATK)
\begin{itemize}
\item Germline mode
\item Somatic mode
\end{itemize}
\item Oncotator (GATK)
\end{enumerate}

\end{frame}

\end{document}
